### ISSUE TITLE
[JENKINS-11256] Maven3 parallel build fails with java.util.ConcurrentModificationException in Jenkins

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-14#4773]
<p>linux, jenkins 1.433, maven 3.0.3, maven project with maven 3 and -T 8 running on a remote ssh slave<br/>
build fails with</p>

<p>&#8212; maven-install-plugin:2.3.1:install (default-install) @ xxxxxxxx &#8212;<br/>
Waiting for Jenkins to finish collecting data<br/>
mavenExecutionResult exceptions not empty<br/>
message : Internal error: java.util.ConcurrentModificationException<br/>
cause : null<br/>
Stack trace : <br/>
org.apache.maven.InternalErrorException: Internal error: java.util.ConcurrentModificationException<br/>
	at org.apache.maven.lifecycle.internal.BuilderCommon.handleBuildError(BuilderCommon.java:128)<span class="error">&#91;INFO&#93;</span> Installing xxxxxxxx.jar to xxxxxxxx</p>

<p>	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:95)<br/>
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:167)<br/>
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:164)<br/>
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:269)<br/>
	at java.util.concurrent.FutureTask.run(FutureTask.java:123)<span class="error">&#91;INFO&#93;</span> Compiling 1 source file to xxxxxxxx</p>

<p>	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:651)<br/>
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:676)<br/>
	at java.lang.Thread.run(Thread.java:595)<br/>
Caused by: java.util.ConcurrentModificationException<br/>
	at java.util.ArrayList.writeObject(ArrayList.java:572)<br/>
	at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)<br/>
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)<br/>
	at java.lang.reflect.Method.invoke(Method.java:592)<br/>
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)<br/>
	at java.util.ArrayList.writeObject(ArrayList.java:569)<br/>
	at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)<br/>
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)<br/>
	at java.lang.reflect.Method.invoke(Method.java:592)<br/>
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)<br/>
	at java.util.HashMap.writeObject(HashMap.java:1039)<br/>
	at sun.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)<br/>
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)<br/>
	at java.lang.reflect.Method.invoke(Method.java:592)<br/>
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)<br/>
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)<br/>
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)<br/>
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)<br/>
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)<br/>
	at hudson.remoting.UserRequest._serialize(UserRequest.java:155)<br/>
	at hudson.remoting.UserRequest.serialize(UserRequest.java:164)<br/>
	at hudson.remoting.UserRequest.<init>(UserRequest.java:62)<br/>
	at hudson.remoting.Channel.callAsync(Channel.java:708)<br/>
	at hudson.maven.Maven3Builder$FilterImpl.executeAsync(Maven3Builder.java:229)<br/>
	at hudson.maven.reporters.MavenFingerprinter.postBuild(MavenFingerprinter.java:96)<br/>
	at hudson.maven.Maven3Builder$MavenExecutionListener.recordProjectSucceeded(Maven3Builder.java:431)<br/>
	at hudson.maven.Maven3Builder$MavenExecutionListener.projectSucceeded(Maven3Builder.java:405)<br/>
	at org.apache.maven.lifecycle.internal.DefaultExecutionEventCatapult.fire(DefaultExecutionEventCatapult.java:74)<br/>
	at org.apache.maven.lifecycle.internal.DefaultExecutionEventCatapult.fire(DefaultExecutionEventCatapult.java:42)<br/>
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:91)<br/>
	... 7 more<br/>
<span class="error">&#91;INFO&#93;</span> <br/>
<span class="error">&#91;INFO&#93;</span> &#8212; maven-compiler-plugin:2.3.2:testCompile (default-testCompile) @ xxxxxxxx &#8212;<br/>
channel stopped<br/>
Archiviere Artefakte<br/>
Finished: FAILURE</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kisa">kisa</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-11256" target="_blank">Maven3 parallel build fails with java.util.ConcurrentModificationException in Jenkins</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.png" title="kutzi's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kutzi">kutzi</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-11-18T14:20:29+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 8
<li><b>imported</b>: 2025-11-14
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>linux, jenkins 1.433, maven 3.0.3, maven project with maven 3 and -T 8 running on a remote ssh slave
build fails with</p>

<p>&#8212; maven-install-plugin:2.3.1:install (default-install) @ xxxxxxxx &#8212;
Waiting for Jenkins to finish collecting data
mavenExecutionResult exceptions not empty
message : Internal error: java.util.ConcurrentModificationException
cause : null
Stack trace : 
org.apache.maven.InternalErrorException: Internal error: java.util.ConcurrentModificationException
	at org.apache.maven.lifecycle.internal.BuilderCommon.handleBuildError(BuilderCommon.java:128)<span class="error">&#91;INFO&#93;</span> Installing xxxxxxxx.jar to xxxxxxxx</p>

<p>	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:95)
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:167)
	at org.apache.maven.lifecycle.internal.LifecycleThreadedBuilder$1.call(LifecycleThreadedBuilder.java:164)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:269)
	at java.util.concurrent.FutureTask.run(FutureTask.java:123)<span class="error">&#91;INFO&#93;</span> Compiling 1 source file to xxxxxxxx</p>

<p>	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:651)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:676)
	at java.lang.Thread.run(Thread.java:595)
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList.writeObject(ArrayList.java:572)
	at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:592)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)
	at java.util.ArrayList.writeObject(ArrayList.java:569)
	at sun.reflect.GeneratedMethodAccessor15.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:592)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)
	at java.util.HashMap.writeObject(HashMap.java:1039)
	at sun.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
	at java.lang.reflect.Method.invoke(Method.java:592)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:917)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1339)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1375)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1347)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1290)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1079)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:302)
	at hudson.remoting.UserRequest._serialize(UserRequest.java:155)
	at hudson.remoting.UserRequest.serialize(UserRequest.java:164)
	at hudson.remoting.UserRequest.&lt;init&gt;(UserRequest.java:62)
	at hudson.remoting.Channel.callAsync(Channel.java:708)
	at hudson.maven.Maven3Builder$FilterImpl.executeAsync(Maven3Builder.java:229)
	at hudson.maven.reporters.MavenFingerprinter.postBuild(MavenFingerprinter.java:96)
	at hudson.maven.Maven3Builder$MavenExecutionListener.recordProjectSucceeded(Maven3Builder.java:431)
	at hudson.maven.Maven3Builder$MavenExecutionListener.projectSucceeded(Maven3Builder.java:405)
	at org.apache.maven.lifecycle.internal.DefaultExecutionEventCatapult.fire(DefaultExecutionEventCatapult.java:74)
	at org.apache.maven.lifecycle.internal.DefaultExecutionEventCatapult.fire(DefaultExecutionEventCatapult.java:42)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:91)
	... 7 more
<span class="error">&#91;INFO&#93;</span> 
<span class="error">&#91;INFO&#93;</span> &#8212; maven-compiler-plugin:2.3.2:testCompile (default-testCompile) @ xxxxxxxx &#8212;
channel stopped
Archiviere Artefakte
Finished: FAILURE</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>linux, jenkins 1.433, maven 3.0.3, maven project with maven 3 and -T 8 running on a remote ssh slave</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.png" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153723&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153723">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Does it happen frequently or only once?</p>
</pre>
</details>

</details>
<p>Does it happen frequently or only once?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>kisa's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153724&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153724">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Every 10-15 builds ... bad enough</p>
</pre>
</details>

</details>
<p>Every 10-15 builds ... bad enough</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/paux.png" title="paux's avatar" /> paux's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153776&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153776">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>All three collections in Maven Fingerprinter need to be synchronized (Collections.synchronized...()) schould be sufficient. This problem could occur at several places.</p>

<p>Also, check wether a given file has alread been digested should be converted into a onestep checkAndSet operation.</p>

<p>I will create a pull request for this.</p>
</pre>
</details>

</details>
<p>All three collections in Maven Fingerprinter need to be synchronized (Collections.synchronized...()) schould be sufficient. This problem could occur at several places.</p>

<p>Also, check wether a given file has alread been digested should be converted into a onestep checkAndSet operation.</p>

<p>I will create a pull request for this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/paux.png" title="paux's avatar" /> paux's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153778&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153778">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Created <a href="https://github.com/pauxus/jenkins/commit/f63bdf359b348707e35d3333a2da22d23397530e" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/pauxus/jenkins/commit/f63bdf359b348707e35d3333a2da22d23397530e</a>, however it landed in the open pull request concerning Parent POM fingerprinting.</p>

<p>Kutzi, could you look at both of them together or should I split it out again?</p>
</pre>
</details>

</details>
<p>Created <a href="https://github.com/pauxus/jenkins/commit/f63bdf359b348707e35d3333a2da22d23397530e" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/pauxus/jenkins/commit/f63bdf359b348707e35d3333a2da22d23397530e</a>, however it landed in the open pull request concerning Parent POM fingerprinting.</p>

<p>Kutzi, could you look at both of them together or should I split it out again?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.png" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153781&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153781">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Thanks for the pull request.
Yes, please create a separate pull request for each issue.</p>

<p>Also, I wonder if the 3 maps/sets are really the problem, as they're (as far as I understand it, are per Maven module and therefore only written in a single thread). My suspicion was much more regarding the FingerprintAction which is added and modified in a non-threadsafe fashion. Have you tried if your change fixes the problem?</p>

<p>I think I'd rather put a synchronized block inside the BuildCallable in postBuild()</p>
</pre>
</details>

</details>
<p>Thanks for the pull request.<br/>
Yes, please create a separate pull request for each issue.</p>

<p>Also, I wonder if the 3 maps/sets are really the problem, as they're (as far as I understand it, are per Maven module and therefore only written in a single thread). My suspicion was much more regarding the FingerprintAction which is added and modified in a non-threadsafe fashion. Have you tried if your change fixes the problem?</p>

<p>I think I'd rather put a synchronized block inside the BuildCallable in postBuild()</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/paux.png" title="paux's avatar" /> paux's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153786&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153786">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Old problem with race conditions - if you need them, they don't show up.</p>

<p>I removed the commit from the pull request - an rechecked again. In my first test, both fingerprinter had very similar hashcodes leading me to the wrong idea that they are reused per module.</p>

<p>As far as I see, the problem occurs during serialization of the BuildCallable itself. Therefore a synchronized block inside BuildCallable should make no difference (Am I overlooking something).</p>

<p>Which leads back to the question: where does ist happen.</p>

<p>The Maven3Builder is shared between modules, it could be the culprit. The stacktrace seems to point to a HashMap containing ArrayLists, with the ArrayList being modified. The ony qualifiying map that I have found would be Maven3Builder.reporters (which would match the stack trace)</p>

<p>However, I did not found anyone modifying one of the valuelists.</p>

<p>Changing MavenModule.createReporters to return a synchronized List should solve the problem, however, I don't like not knowing where the problem actually arises.</p>

<p>I will further look into this.</p>

<p>UPDATE:</p>

<p>Its an ArrayList in an ArrayList in the reporters map could be some plugin.</p>
</pre>
</details>

</details>
<p>Old problem with race conditions - if you need them, they don't show up.</p>

<p>I removed the commit from the pull request - an rechecked again. In my first test, both fingerprinter had very similar hashcodes leading me to the wrong idea that they are reused per module.</p>

<p>As far as I see, the problem occurs during serialization of the BuildCallable itself. Therefore a synchronized block inside BuildCallable should make no difference (Am I overlooking something).</p>

<p>Which leads back to the question: where does ist happen.</p>

<p>The Maven3Builder is shared between modules, it could be the culprit. The stacktrace seems to point to a HashMap containing ArrayLists, with the ArrayList being modified. The ony qualifiying map that I have found would be Maven3Builder.reporters (which would match the stack trace)</p>

<p>However, I did not found anyone modifying one of the valuelists.</p>

<p>Changing MavenModule.createReporters to return a synchronized List should solve the problem, however, I don't like not knowing where the problem actually arises.</p>

<p>I will further look into this.</p>

<p>UPDATE:</p>

<p>Its an ArrayList in an ArrayList in the reporters map could be some plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.png" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=153946&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-153946">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>According to Olivier this might be a bug in Maven 3.0.3:
<a href="http://jenkins.361315.n4.nabble.com/Maven2-3-Parallel-Builds-tp3896403p3904271.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Maven2-3-Parallel-Builds-tp3896403p3904271.html</a></p>
</pre>
</details>

</details>
<p>According to Olivier this might be a bug in Maven 3.0.3:<br/>
<a href="http://jenkins.361315.n4.nabble.com/Maven2-3-Parallel-Builds-tp3896403p3904271.html" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins.361315.n4.nabble.com/Maven2-3-Parallel-Builds-tp3896403p3904271.html</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/paux.png" title="paux's avatar" /> paux's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=155082&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155082">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>Short update:</p>

<p>I could not reproduce the problem using a test installation, I will try another test using the exact plugin configuration of the actual system.</p>
</pre>
</details>

</details>
<p>Short update:</p>

<p>I could not reproduce the problem using a test installation, I will try another test using the exact plugin configuration of the actual system.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/paux.png" title="paux's avatar" /> paux's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=155553&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155553">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>@kutzi: My development environment seems broken. Could you do that change?</p>

<p>Thx</p>
</pre>
</details>

</details>
<p>@kutzi: My development environment seems broken. Could you do that change?</p>

<p>Thx</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kutzi.png" title="kutzi's avatar" /> kutzi's <a href="https://issues.jenkins.io/browse/JENKINS-11256?focusedId=155574&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155574">comment</a>:</i></summary>


<details><summary><sub><i>Raw content of original comment:</i></sub></summary>

<pre>
<p>While it might make sense to make TestResult thread-safe, I'd also ask if it's actually needed that SurefireArchiver#result is serialized.
Maybe that could be changed, too?</p>
</pre>
</details>

</details>
<p>While it might make sense to make TestResult thread-safe, I'd also ask if it's actually needed that SurefireArchiver#result is serialized.<br/>
Maybe that could be changed, too?</p>
